internal/generic/generic.go

```diff
@@ -0,0 +1,154 @@
+/*
+Package generic provides a set of generic helper functions,
+primarily aimed at common channel operations.
+*/
+package generic
+
+import (
+	"context"
+	"sync"
+)
+
+func ctxRange[in any](ctx context.Context, input <-chan in) <-chan in {
+	relay := make(chan in, cap(input))
+	go func() {
+		defer close(relay)
+		for {
+			select {
+			case element, ok := <-input:
+				if !ok {
+					return
+				}
+				select {
+				case relay <- element:
+				case <-ctx.Done():
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return relay
+}
+
+func totalBuff[in any](inputs []<-chan in) (total int) {
+	for _, ch := range inputs {
+		total += cap(ch)
+	}
+	return
+}
+
+func CtxJoin[in any](ctx context.Context, inputs ...<-chan in) <-chan in {
+	out := make(chan in, totalBuff(inputs))
+	go func() {
+		defer close(out)
+		for _, source := range inputs {
+			for element := range ctxRange(ctx, source) {
+				select {
+				case out <- element:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+	return out
+}
+
+func CtxMerge[in any](ctx context.Context, sources ...<-chan in) <-chan in {
+	var (
+		mergedWg  sync.WaitGroup
+		mergedCh  = make(chan in, totalBuff(sources))
+		mergeFrom = func(ch <-chan in) {
+			defer mergedWg.Done()
+			for value := range ctxRange(ctx, ch) {
+				select {
+				case mergedCh <- value:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	)
+
+	mergedWg.Add(len(sources))
+	for _, source := range sources {
+		go mergeFrom(source)
+	}
+
+	go func() { mergedWg.Wait(); close(mergedCh) }()
+
+	return mergedCh
+}
+
+func CtxTakeAndCancel[in any](ctx context.Context, cancel context.CancelFunc,
+	inputs <-chan in, count int,
+) <-chan in {
+	relay := make(chan in, count)
+	go func() {
+		defer close(relay)
+		defer cancel()
+		for element := range ctxRange(ctx, inputs) {
+			if count == 0 {
+				return
+			}
+			select {
+			case relay <- element:
+			case <-ctx.Done():
+				return
+			}
+			count--
+		}
+	}()
+	return relay
+}
+
+func ForEachOrError[in any](ctx context.Context,
```

Comment:
While looking at this functions use in the `staging/mount` branch, 11/15 callers pass a `nil` error channel since they don't have/need one.
I'm thinking we should instead export `ctxRange` and eliminate `ForEachOrError`.

For the remaining calls that do have an `errs` chan, we could expose some kind of function|chan wrapper.
Something like: `newWrapper(ctx, fn, errs) fnWrapped` which calls `fn` and checks `errs` in some way.
Need to think about it.

Reply:
Is there a resource usage improvement with the wrapper route? I feel like it's worth it to remove an argument if isn't always needed.

Reply:
>Is there a resource usage improvement with the wrapper route? 

I doubt it (but didn't check).
It'd be 1 less `select` `case`, but the overhead of the wrapper would likely be a net negative in every way (when required).
(Bad idea on my part, was overthinking it)

`ctxRange` combined with a `range` serves the same purpose as calling `ForEachOrError` without an `errs` chan.
E.g.
```go
phase := func(element) error {
    useElem(element)
    ...
}
ForEachOrError(ctx, input, nil, phase)
```
can turn into
```go
for element := range ctxRange(ctx, input) { 
    useElem(element)
    ...
```
which seems a lot less magic.

***

For callers that do have to `select` on `errs`, we can either leave `ForEachOrError` as-is or eliminate it and rely on the caller to do the select themselves at the call-site.

```go
phase := func(element) error {
    useElem(element)
    ...
}
ForEachOrError(ctx, input, errs, phase){
```
has to explode into at least

```go
for input != nil ||
    errs != nil {
    select {
    case element, ok := <-input:
        if !ok {
            input = nil
            continue
        }
        useElem(element)
        ...
    case err, ok := <-errs:
        if !ok {
            errs = nil
            continue
        }
        handle(err)
    case <-ctx.Done():
        return
    }
```
`ForEachOrError` does a good job of hiding the `select` boilerplate, but it might be worth losing the golf points for the clarity at the callsite.
I'm not sure yet.
We might leave it out of this patchset since it's only called once in the refactor I have pending. And if we use this pattern again, we can add it back.

I'll consider if there's another alternative as well. Wrapping the `input` and `errs` channels to form a results/tuple channel comes to mind, but that adds overhead that currently isn't necessary. (each input would need to be relayed, along with mostly nil errors within a weird generic struct)

Reply:
We discussed this and ended up eliminating `ForEachOrError` and `ctxRange`.
For now, we just assure that all sends are cancelable and rely on that in conjuncture with Go's builtin`for range`+`select`.
Literally: `for value := range values { useValue(value) }` where values uses a generator which uses a context (and the range body selects on any of its channel operations).

For code which had to deal with both value and error channels, we used the last approach (in the previous post,) via a [generic Tuple (Couple) type](https://github.com/djdv/go-filesystem-utils/blob/f2dbe83f6d3fa21206f1cf658491ae6fe2921c28/internal/generic/generic.go#L9).
Which can be used to construct couplings of any 2 types (including tuples recursively).

We can use this to make generic generators like [CtxEither](https://github.com/djdv/go-filesystem-utils/blob/f2dbe83f6d3fa21206f1cf658491ae6fe2921c28/internal/generic/generic.go#L17) and use it to write code like this:
```go
for valueOrErr := range generic.CtxEither(ctx, values, errs) {
    if err := valueOrErr.Right; err != nil {
        return nil, err
    }
    value := valueOrErr.Left
    useValue(value)
}
```
Without having to think too much about channel (receive and merge) operations on the receiver side.

It's also useful for eliminating type specific tuples (distinc 2n results and pair struct types) and their helper functions, in favor of aliases to generic couples.
```go
// Shorthand type alias for a pair of 2 types.
type fieldParam = generic.Couple[reflect.StructField, parameters.Parameter]
```
```go
// Joins 2 input channels into a single Couple channel (aliased).
func generateFieldParams(ctx, fields, params) <-chan fieldParam{
    out := make(chan fieldParam)
    out <- fieldParam{Left: field, Right: param} 
}
```
```go
// Uses the same Couple type with different semantics.
// Combining an "Either/Or" Couple and a "Both/And" Couple.
// Where Both is a field and a parm, and either is a Both or an error.
for pairOrErr := range generic.CtxEither(subCtx, fieldParams, errs) {
	if err := pairOrErr.Right; err != nil {
		return nil, err
	}
	var (
		fieldAndParam = pairOrErr.Left
		field         = fieldAndParam.Left
		param         = fieldAndParam.Right
		opt, err      = newSettingsOption(field, param, makers)
	)
```